### PR TITLE
feat: finer dependency analysis between libraries (#4572)

### LINKED ISSUE #4572: Finer dependency analysis between libraries
Currently, `dune`  uses per-module dependency information (as reported by `ocamldep`) *within* a given library. Between libraries, `dune` simply overapproximates by assuming that if a library `libA` depends on a library `libB` then every module of `libA` depends on every module of `libB`.

This means that as soon as a module in `libB` is changed, then every module in `libA` is recompiled (this is somewhat aliviated in `dev` mode, where changes to *implementations* in `libB` will not cause recompilation of of modules in `libA`).

This causes extra recompilation (especially in the case of unwrapped libraries). Today in the dev meeting it was proposed to use the output of `ocamldep` to specify per-module dependencies not only within a given library but also across libraries.

In the case when `libB` is wrapped, there is less potential improvement (given that as soon as a module in `libA` depends on the interface module of `libB` one must assume that it could depend on any module of `libB`), but even in that case, if a module in `libA` does not depend at all in `libB` one could avoid some recompilation.

cc @rgrinberg

### PR DESCRIPTION
**Pull request by robinbb**
Implements per-module library dependency filtering as proposed in #4572. When
library A depends on library B, Dune currently assumes every module in A
depends on every module in B — causing full recompilation of A whenever any
module in B changes. This PR uses ocamldep's per-module output to declare
cross-library file dependencies only for libraries a module actually
references.

This PR builds on prerequisite PRs:
- #14017 — baseline tests documenting current recompilation behavior (merged)
- #14031 — test documenting module name shadowing between stanzas and libraries
- #14100 — test verifying library file deps in compilation rules and sandboxed builds
- #14101 — safety test for incremental builds with transparent aliases
- #14129 — test verifying incremental builds with alias re-exported libraries
- #14030 — refactor moving `Hidden_deps` from `Includes.make` to per-module in `build_cm`

Depends on #14030 and #14129.

The changes in this PR (on top of the above):
- A new `Lib_index` (in `lib_file_deps.ml`) maps library entry module names
  back to their libraries. It is computed once per stanza and stored in
  `Compilation_context.t`.
- `read_immediate_deps_raw_of` reads raw ocamldep output (including
  cross-library references that were previously silently dropped), filters
  out stanza-internal names, and looks up the remainder in the index to
  determine which libraries are actually needed.
- Conservative fallbacks ensure correctness: special module kinds, link-time
  generated modules, and unresolvable references fall back to all-library
  deps.
- Baseline test expectations are updated to show the improved behaviour.

Resolves #4572

### CODE REVIEW

**Comment by rgrinberg** [comment]
Thanks for working on this. I expect users are going to be quite pleased to have this feature. Some preliminary questions:

> Includes.make no longer bundles Hidden_deps with -I flags. Include
flags remain shared across all modules in a stanza; file dependencies are
now declared per-module in build_cm.

Why are the include flags are needed if we don't expect the library to be used by the module?

> For each module, read_immediate_deps_raw_of reads the raw ocamldep output
(including cross-library references that were previously silently dropped),
filters out stanza-internal names, and looks up the remainder in the index
to determine which libraries are actually needed.

Do we have tests for when an internal module name is shadowing a library?

> Conservative fallbacks ensure correctness: alias and wrapped-compat modules
skip library deps entirely; special module kinds, link-time generated
modules, and unresolvable references fall back to all-library deps.

Is it possible to move these changes to a separate PR and rebase this work on top of that?

**Comment by robinbb** [comment]
> Is it possible to move these changes to a separate PR and rebase this work on top of that?

@rgrinberg I have tried to do so with this PR https://github.com/ocaml/dune/pull/14030, depending on what you really meant by "these changes". :-)

**Comment by robinbb** [comment]
> Thanks for working on this. I expect users are going to be quite pleased to have this feature.

You're very welcome, and thank you for your prompt attention to it!

**Comment by robinbb** [comment]
> Do we have tests for when an internal module name is shadowing a library?

I guess not. I have made #14031 for this. Please have a look.

File: src/dune_rules/module_compilation.ml [hunk 1]
```diff
@@ -286,6 +286,87 @@ let build_cm
...
+            let used_entries =
+              if references_root_module
+              then all_entries
+              else (
+                let used =
+                  Lib_file_deps.Lib_index.filter_libs
+                    lib_index
+                    ~referenced_modules:external_modules
+                in
+                if List.is_empty used && not (Module_name.Set.is_empty external_modules)
```
Comment (Alizter) [inline src/dune_rules/module_compilation.ml]: Do `stdlib` modules get reported as external_modules here? If so, doesn't this mean any stdlib modules like `Printf` will trigger the fallback behaviour?

This would mean that any module that uses the stdlib would be affected. Could you add a test case that demonstrates this is (or is not) the case?

Comment (robinbb) [reply src/dune_rules/module_compilation.ml]: Indeed I have modified #14017 to add a test for this.

**Comment by nojb** [comment]
Thanks for working on this patch @robinbb!

I tried this in our codebase by adding a new (unused) module to a library which is close to the root of the dependency tree, and doing `dune build`. Then I modified the unused module and redid `dune build`. At this point I expected for there not to be any recompilation at all, since the modified module is not used anywhere in the codebase. However, I observe some recompilation of modules (in other libraries) which do not depend on it. The libraries involved are all unwrapped.

Do you have any insight as to what may be causing the extra recompilation?

**Comment by nojb** [comment]
> Do you have any insight as to what may be causing the extra recompilation?

I think one case where unexpected extra recompilation is triggered is when there a library consisting of a single module. If memory serves there is an optimization that avoids calling `ocamldep` in this case, but that may be defeating the optimization in this patch.

**Comment by Alizter** [comment]
Transparent module aliases hide library references from `ocamldep` when a module re-exports a library wrapper via `module M = Mylib`, `ocamldep` will miss the dependency on `Mylib`.  

Here is the regression:
```

  $ cat > dune-project << EOF
  > (lang dune 3.0)
  > EOF
  $ mkdir lib && cat > lib/dune << EOF
  > (library (name mylib))
  > EOF
  $ cat > lib/mylib.ml << EOF
  > let v = 42
  > EOF
  $ cat > lib/mylib.mli << EOF
  > val v : int
  > EOF
  $ cat > dune << EOF
  > (executable (name main) (libraries mylib))
  > EOF
  $ cat > re.ml << EOF
  > module M = Mylib
  > EOF
  $ cat > re.mli << EOF
  > module M = Mylib
  > EOF
  $ cat > main.ml << EOF
  > let () = print_int Re.M.v
  > EOF

  $ dune build ./main.exe

Change mylib's interface:

  $ cat > lib/mylib.mli << EOF
  > val v : int
  > val w : string
  > EOF
  $ cat > lib/mylib.ml << EOF
  > let v = 42
  > let w = ""
  > EOF

The incremental build must succeed:

  $ dune build ./main.exe
  File "_none_", line 1:
  Error: Files .main.eobjs/native/dune__exe__Main.cmx and lib/mylib.cmxa
         make inconsistent assumptions over interface Mylib
  [1]
```

This probably also occurs for any module from `Mylib` not just the wrapper, so affects unwrapped, wrapped and singleton buildable stanzas. Before this change, the coarse glob would handle this unknowingly.

**Comment by nojb** [comment]
> > Do you have any insight as to what may be causing the extra recompilation?
> 
> I think one case where unexpected extra recompilation is triggered is when there a library consisting of a single module. If memory serves there is an optimization that avoids calling `ocamldep` in this case, but that may be defeating the optimization in this patch.

Continuing to investigate this, an `.mli` in a library `A` containing only
```
val foo: Buffer.t
```
triggers the recompilation of the `.cmi` every time when a module in a library on which `A` depends on is changed. From a distance this looks like some missing logic to handle dependencies on stdlib modules.

**Comment by robinbb** [comment]
Thanks @Alizter for the transparent alias regression report and test case. The
scenario you described (and the exact test) is now included in the branch and
passes.

The root cause: `main.ml` uses `Re.M.v` where `re.ml` has `module M = Mylib`.
`ocamldep` on `main.ml` reports only `Re`, not `Mylib`. The original per-module
filtering saw no external references for `main.ml` and declared zero library
deps, causing stale builds.

The fix: `build_cm` now unions external module references across the transitive
intra-stanza dependency closure (via `Dep_graph.top_closed_implementations`),
not just the module's own `ocamldep` output. Since `re.ml`'s `ocamldep`
includes `Mylib`, the union picks it up and `main.ml` correctly depends on
`Mylib`'s files.

During this investigation I also found and fixed a separate issue:
`requires_compile` for stanzas that depend on virtual library implementations
does not include the virtual library itself. This meant the `Lib_index` was
structurally incomplete, and even glob-dep fallbacks could not cover the
virtual library's modules. Filtering is now disabled for stanzas whose requires
include virtual library implementations. This fixes a clean-build failure in
`virtual-libraries/unwrapped.t` that was present in the original version of
this PR.

**Comment by robinbb** [comment]
@nojb I wonder if https://github.com/ocaml/dune/pull/14079 has all of the fixes that you might be looking for? Care to give it a try on your code base, possibly with the modifications that mention trying, above?

**Comment by nojb** [comment]
> @nojb I wonder if #14079 has all of the fixes that you might be looking for? Care to give it a try on your code base, possibly with the modifications that mention trying, above?

Yes, it seems to fix the issues reported above. Thanks!

While testing, I found another issue that seems to be related to this patch, when doing `make dev-switch` directly on this branch, a compilation failure while building Menhir:

image

**Comment by robinbb** [comment]
~@nojb I am seeing a menhir-related failure on the CI of the related PR, #14030 , but not on this PR which includes #14030. For that reason, I think it has something to do with the opam invocation and the current menhir. I think that the CI build step that is failing probably doesn't actually use the Dune that is built using the code in this PR. (The build step is called "install deps", and seems to use opam.) Can anyone enlighten me, or confirm?~ The problem is in the refactor PR, upon which this one (and others) are based. I am correcting it, now. Thanks so much for your help, @nojb !

**Comment by Alizter** [comment]
There are changes in the following tests:
- `test/blackbox-tests/test-cases/reporting-of-cycles.t`
- `test/blackbox-tests/test-cases/private-modules/accessible-via-public.t`

The first is a cycle that we are no longer reporting, we need to decide if this is because the test was relying on the old dep behaviour or something else is broken (likely the former).

The second seems like a regression with private modules.

**Comment by rgrinberg** [comment]
Can we also run a meaningful revdep check on this? This has the potential to affect a bunch of downstream packages.

**Comment by Khady** [comment]
Very exciting work, thanks!

I've tried against our monorepo, but unfortunately a bunch of code doesn't compile. I can't really extract the whole thing, but from what I understand:
-  `Unbound module XXX`, with XXX being a module in a library that is `wrapped false`
- `Error: The module Fmt_network is an alias for module Ahrefscore__Fmt_network, which is missing`, with fmt_network.ml being generated by a hand written rule
- some `open Xxx` are failing, but I don't really know why yet
```
File "backend/foo.mli", line 2, characters 5-13:
2 | open Bar_types
         ^^^^^^^^
Error: This is an alias for module Ahrefscore__Bar_types, which is missing
```

**Comment by robinbb** [comment]
The `@package-cycle` change is expected. Libraries `a.1` and `b` declare mutual
dependencies, but their `.ml` files are empty — no module actually references
anything from the other library. With per-module filtering, empty modules
produce no library file deps, so the build graph edge that created the cycle
disappears.

This is correct: the cycle was purely structural (declared `(libraries ...)`
but no code reference). The test expectation should be updated to remove the
`@package-cycle` error. The other cycle tests (simple-repro, x1, indirect-deps)
are unaffected because they involve actual code-level or rule-level cycles.

**Comment by robinbb** [comment]
The private modules regression (`accessible-via-public.t`) is fixed. The
`accessible-via-public.t` test now passes unchanged. There is no longer any
change to the private modules test output.

**Comment by robinbb** [comment]
@Khady I'm very glad that you invested the time in trying out this build. (Note that the PR was and is still in draft mode, so beware of investing too much time in it.) I think that I have addressed the `wrapped false` problem that you mentioned, in a subsequent version of this branch. I hope that it works for you.

File: test/blackbox-tests/test-cases/reporting-of-cycles.t/run.t [hunk 2]
```diff
@@ -5,12 +5,6 @@ start running things. In the past, the error was only reported during
 the second run of dune.
 
   $ dune build @package-cycle
-  Error: Dependency cycle between:
```
Comment (Alizter) [inline test/blackbox-tests/test-cases/reporting-of-cycles.t/run.t:8]: The removal of this error is strange, because it means we will be able to build packages together in a workspace that otherwise would have cycle errors and cause the (opam and dune) solvers to fail to find a solution.

I think we need to do cycle detection at the level of packages in order to avoid such situations. Before it wasn't needed because the dependencies were coarse enough to imply the sufficient conditions to avoid a cycle.

@rgrinberg any advice on where and how this detection should be done?

**Comment by Leonidas-from-XIV** [comment]
> Can we also run a meaningful revdep check on this? This has the potential to affect a bunch of downstream packages.

Yes, we have a (manually triggered) workflow for this, so provided the branch is rebased on a reasonably recent version of `main` (I fixed some smaller issues in it recently) it should be possible to trigger [this workflow](https://github.com/ocaml/dune/actions/workflows/revdeps.yml) by

  * picking @robinbb's branch from the dropdown
  * setting the list of packages to some meaningful set (or `all` but [it takes a long time, testing it at time of writing](https://github.com/ocaml/dune/actions/runs/24181111885/job/70574041011))
  * keeping the Dune version as `dev` (which means it will take the version of the branch itself, not `main`)

For smaller package sets this takes about 10 minutes, bigger sets might take several hours I assume.

**Comment by robinbb** [comment]
I am chasing down a CI failure in this branch. The failure is good, in the sense that it has caught legitimate regressions introduced by prior versions of this branch - ones that were not caught by the cram test suite.

**Comment by art-w** [comment]
Very cool work indeed!

Is you new PR https://github.com/ocaml/dune/pull/14116 superseding this one?

Either way, I think it's worth spending more time to improve the test suite, because it's easy to accidentally rely on subtle behaviors that are more permissive than the general case. For example, the `-opaque` flag makes it possible to skip some recompilations on benign changes, even though these recompilations would normally be required.

Here's a small reproduction which should help debug the issue where a library re-exports one of its dependencies with a module alias. This cram test also shows the subtleties in the setup that could make it pass (when it shouldn't!)

---

First we need a library where we'll perform the changes:

    $ mkdir impl
    $ cat > impl/dune <<EOF
    > (library (name impl))
    > EOF
    $ cat > impl/impl.ml <<EOF
    > let foo = "initial build"
    > EOF

Then another library, which exposes an alias to `impl`:

    $ mkdir alias
    $ cat > alias/dune <<EOF
    > (library (name alias) (libraries impl))
    > EOF
    $ cat > alias/alias.ml <<EOF
    > module Impl = Impl
    > EOF

And a binary which depends on `Alias` to access `Impl`:

    $ mkdir bin
    $ cat > bin/dune <<EOF
    > (executable (name main) (libraries alias))
    > EOF
    $ cat > bin/main.ml <<EOF
    > let () = print_endline Alias.Impl.foo
    > EOF

Compilation is different when the executable is defined in a single file, so the bug doesn't trigger unless we also add an empty unused file: (!!)

    $ touch bin/unused.ml

The first build succeeds:

    $ dune exec ./bin/main.exe
    initial build

Now we introduce a "soft" update, such that `impl.cmi` isn't modified:

    $ cat > impl/impl.ml <<EOF
    > let foo = "second build, no change to cmi"
    > EOF

In that specific case, we are allowed to not recompile `bin/main.ml` because the `-opaque` flag guarantees we are not dependent on the implementation of `Impl`: (the linker checks that the cmi digest hasn't changed)

    $ dune exec ./bin/main.exe
    second build, no change to cmi

But if the `cmi` of `Impl` changed, then skipping the recompilation of `main.ml` is wrong: (the linker check fails)

    $ cat > impl/impl.ml <<EOF
    > let new_value = 42
    > let foo = "third build, forced a cmi update"
    > EOF

    $ dune exec ./bin/main.exe
    File "_none_", line 1:
    Error: Files bin/.main.eobjs/native/dune__exe__Main.cmx and impl/impl.cmxa
           make inconsistent assumptions over interface Impl
    [1]

---

Regarding revdeps testing, one issue to keep in mind is that testing a clean build (and not an incremental build) is likely to hide the failures:

    $ dune clean
    $ dune exec ./bin/main.exe
    third build, forced a cmi update

**Comment by robinbb** [comment]
> Is you new PR #14116 superseding this one?

No, not necessarily. I have 2 PRs because I need to test more than one thing at a time, and I need access to the CI invocations. (Very difficult to replicate the CI corner cases that are failing, now, for this effort. Non-corner cases are passing.)

**Comment by robinbb** [comment]
> Either way, I think it's worth spending more time to improve the test suite, because it's easy to accidentally rely on subtle behaviors that are more permissive than the general case. For example, the `-opaque` flag makes it possible to skip some recompilations on benign changes, even though these recompilations would normally be required.

Yes, agreed - we need more tests. The absence of test coverage is what is making it so difficult to get this right. It is difficult to see just how many different scenarios are actually working (but non-optimal), in the status quo. I have submitted many augmentations already to the test suite in dependency PRs to this one (14017, 14031, 14100, 14101, 14103) in an effort to reduce the risk, already. More still are required.

I see your advice for a suggested new test(s). I'll try to recreate that in a cram test. Possibly in a different PR (onto which I will rebase this branch).

**Comment by robinbb** [comment]
> > Is you new PR #14116 superseding this one?
> 
> No, not necessarily. I have 2 PRs because I need to test more than one thing at a time, and I need access to the CI invocations. (Very difficult to replicate the CI corner cases that are failing, now, for this effort. Non-corner cases are passing.)

In the end, yes, it is.

File: test/blackbox-tests/test-cases/reporting-of-cycles.t/run.t (hunk 2 above)
Comment (robinbb) [reply test/blackbox-tests/test-cases/reporting-of-cycles.t/run.t:8]: Moving discussion to https://github.com/ocaml/dune/pull/14116#issuecomment-4227057455

**Comment by robinbb** [comment]
Deprecated in favour of #14116
